File: 2024/day-08/main.ex
defmodule GridAntenna do
  def manhattan_distance({r1, c1}, {r2, c2}) do
    abs(r1 - r2) + abs(c1 - c2)
  end

  def main() do
    infile = System.argv() |> Enum.at(0, "input.txt")
    grid = File.read!(infile) |> String.split("\n", trim: true)

    positions =
      Enum.with_index(grid)
      |> Enum.reduce(%{}, fn {row, r}, acc ->
        row
        |> String.graphemes()
        |> Enum.with_index()
        |> Enum.reduce(acc, fn {ch, c}, acc ->
          if ch != "." do
            Map.update(acc, ch, [{r, c}], &[{r, c} | &1])
          else
            acc
          end
        end)
      end)

    {a1, a2} =
      Enum.reduce(0..(length(grid) - 1), {MapSet.new(), MapSet.new()}, fn r, {a1, a2} ->
        Enum.reduce(0..(String.length(hd(grid)) - 1), {a1, a2}, fn c, {a1, a2} ->
          Enum.reduce(positions, {a1, a2}, fn {_k, pos}, {a1, a2} ->
            Enum.reduce(pos, {a1, a2}, fn {r1, c1}, {a1, a2} ->
              Enum.reduce(pos, {a1, a2}, fn {r2, c2}, {a1, a2} ->
                if {r1, c1} != {r2, c2} do
                  d1 = manhattan_distance({r, c}, {r1, c1})
                  d2 = manhattan_distance({r, c}, {r2, c2})
                  dr1 = r - r1
                  dr2 = r - r2
                  dc1 = c - c1
                  dc2 = c - c2

                  a1 =
                    if (d1 == 2 * d2 or d1 * 2 == d2) and dr1 * dc2 == dc1 * dr2,
                       do: MapSet.put(a1, {r, c}),
                       else: a1

                  a2 =
                    if dr1 * dc2 == dc1 * dr2,
                       do: MapSet.put(a2, {r, c}),
                       else: a2

                  {a1, a2}
                else
                  {a1, a2}
                end
              end)
            end)
          end)
        end)
      end)

    IO.puts(MapSet.size(a1))
    IO.puts(MapSet.size(a2))
  end
end

GridAntenna.main()

File: 2024/day-08/main.rs
use std::collections::{HashMap, HashSet};
use std::fs;

fn manhattan_distance(r1: isize, c1: isize, r2: isize, c2: isize) -> isize {
    (r1 - r2).abs() + (c1 - c2).abs()
}

fn main() {
    let infile = std::env::args().nth(1).unwrap_or("input.txt".to_string());
    let content = fs::read_to_string(&infile).expect("Error reading file");

    let grid: Vec<&str> = content.lines().collect();
    let r = grid.len() as isize;
    let c = grid[0].len() as isize;

    let mut positions: HashMap<char, Vec<(isize, isize)>> = HashMap::new();
    for (row, line) in grid.iter().enumerate() {
        for (col, ch) in line.chars().enumerate() {
            if ch != '.' {
                positions.entry(ch).or_insert_with(Vec::new).push((row as isize, col as isize));
            }
        }
    }

    let mut a1 = HashSet::new();
    let mut a2 = HashSet::new();

    for r in 0..r {
        for c in 0..c {
            for (_ch, pos) in &positions {
                for &(r1, c1) in pos {
                    for &(r2, c2) in pos {
                        if (r1, c1) != (r2, c2) {
                            let d1 = manhattan_distance(r, c, r1, c1);
                            let d2 = manhattan_distance(r, c, r2, c2);
                            let dr1 = r - r1;
                            let dr2 = r - r2;
                            let dc1 = c - c1;
                            let dc2 = c - c2;

                            if (d1 == 2 * d2 || d1 * 2 == d2) && dr1 * dc2 == dc1 * dr2 {
                                a1.insert((r, c));
                            }
                            if dr1 * dc2 == dc1 * dr2 {
                                a2.insert((r, c));
                            }
                        }
                    }
                }
            }
        }
    }

    println!("{}", a1.len());
    println!("{}", a2.len());
}
